### PR TITLE
Add user profile picture

### PR DESCRIPTION
**Pull request by OmeGak**
Partially addresses #936.

This PR adds:
- A file upload field in the user personal details form.
- Display of the profile picture in the user profile dashboard.

## Functional testing instructions

Upload profile pic:
1. Go to user profile settings (`user/<user_id>/profile/`)
2. Upload a file from your computer
3. Verify that the picture is uploaded successfully
4. Verify that the profile picture is displayed in the user profile dashboard (`user/<user_id>/dashboard/`)

Delete profile pic:
1. Go to user profile settings (`user/<user_id>/profile/`)
2. Delete the profile picture
3. Verify that the profile picture is not displayed in the user profile dashboard (`user/<user_id>/dashboard/`)

## Screenshots

image

image

image

### CODE REVIEW

File: indico/modules/users/controllers.py [hunk 1]
```diff
@@ -177,13 +182,63 @@ class RHPersonalData(RHUserBase):
...
+        return WPUser.render_template('personal_data.html', 'personal_data', user=self.user, form=form,
+                                      pic_form=pic_form)
+
+
+class RHProfilePictureDisplay(RHUserBase):
+    def _process(self):
+        if not self.user.has_picture:
+            raise NotFound
+        metadata = self.user.picture_metadata
+        return send_file(metadata['filename'], BytesIO(self.user.picture), mimetype=metadata['content_type'],
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:204]: I don't think we use the original filename here. Imagine a user using a somewhat inappropriate name...

FWIW I think it isn't a good idea to send the original filename for the other images where we do this either (even though it was probably me writing some of those RHs 🙈) - storing them is one thing (and that's nice so the user sees their original filename), but for sending I'd use something generic, like "user.png"

File: indico/modules/users/controllers.py [hunk 2]
```diff
@@ -177,13 +182,63 @@ class RHPersonalData(RHUserBase):
...
+                         conditional=True)
+
+
+class RHProfilePicture(RHUserBase):
+    def _process(self):
+        f = request.files['picture']
+        try:
+            pic = Image.open(f)
+        except IOError:
+            raise UserValueError(_('You cannot upload this file as a room picture.'))
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:214]: room picture?

File: indico/modules/users/controllers.py [hunk 3]
```diff
@@ -177,13 +182,63 @@ class RHPersonalData(RHUserBase):
...
+        if pic.format.lower() not in {'jpeg', 'png'}:
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:215]: what about allowing gif as well?

File: indico/modules/users/models/users.py [hunk 4]
```diff
@@ -252,6 +254,17 @@ class User(PersonMixin, db.Model):
         nullable=False,
         default=lambda: unicode(uuid4())
     )
+    #: the user profile picture
+    picture = db.deferred(db.Column(
+        db.LargeBinary,
+        nullable=True
+    ))
+    # user profile picture metadata
```
Comment (ThiefMaster) [inline indico/modules/users/models/users.py:262]: missing `:` (to make the comment docstring-ish)

File: indico/modules/users/templates/personal_data.html [hunk 5]
```diff
@@ -48,4 +63,29 @@
             </div>
         </div>
     </div>
+
+    <script>
+        $('.pic-form').on('indico:fieldsSaved', function(e, response) {
+            var hidden = !response.content;
```
Comment (ThiefMaster) [inline indico/modules/users/templates/personal_data.html:69]: Let's use `const` (or `let` if necessary) and arrow functions for callbacks (at least when `this` is not needed, so not in this callback) in new code

File: indico/modules/users/controllers.py [hunk 6]
```diff
@@ -177,13 +182,63 @@ class RHPersonalData(RHUserBase):
...
+        content = image_bytes.read()
+        self.user.picture = content
+        self.user.picture_metadata = {
+            'hash': crc32(content),
+            'size': len(content),
+            'filename': os.path.splitext(secure_filename(f.filename, 'logo'))[0] + '.png',
+            'content_type': 'image/png'
+        }
+        flash(_('Profile picture uploaded'), 'success')
+        logger.info("Profile picture of user %s uploaded by %s", self.user, session.user)
```
Comment (mic4ael) [inline indico/modules/users/controllers.py:231]:
```suggestion
        logger.info('Profile picture of user %s uploaded by %s', self.user, session.user)
```

File: indico/modules/users/controllers.py [hunk 7]
```diff
@@ -177,13 +182,63 @@ class RHPersonalData(RHUserBase):
...
+        return jsonify_data(content=get_picture_data(self.user))
+
+
+class RHProfilePictureDelete(RHUserBase):
+    def _process(self):
+        self.user.picture = None
+        self.user.picture_metadata = None
+        flash(_('Profile picture deleted'), 'success')
+        logger.info("Profile picture of user %s deleted by %s", self.user, session.user)
```
Comment (mic4ael) [inline indico/modules/users/controllers.py:240]:
```suggestion
        logger.info('Profile picture of user %s deleted by %s', self.user, session.user)
```

File: indico/web/client/styles/modules/_dashboard.scss [hunk 8]
```diff
@@ -193,10 +193,19 @@
...
+
+            &.placeholder {
+                font-size: 70px;
+            }
+
+            &:not(.placeholder) {
+                height: 140px;
+            }
         }
+
```
Comment (mic4ael) [inline indico/web/client/styles/modules/_dashboard.scss:207]: redundant empty lines?

File: indico/modules/users/controllers.py (hunk 2 above)
Comment (OmeGak) [reply indico/modules/users/controllers.py:214]: Copypasta leftovers.

File: indico/modules/users/controllers.py (hunk 3 above)
Comment (OmeGak) [reply indico/modules/users/controllers.py:214]: Would you really want people to use animated gifs as profiles pictures?

Comment (OmeGak) [reply indico/modules/users/controllers.py:214]: aAYrp7R

Comment (ThiefMaster) [reply indico/modules/users/controllers.py:214]: no but that part goes away when converting to PNG anyway

File: indico/migrations/versions/20200428_1807_532f0ea25bb1_add_user_profile_picture.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    op.add_column('users', sa.Column('picture_metadata', postgresql.JSONB(astext_type=sa.Text()), nullable=False,
+                                     server_default='null'),
+                  schema='users')
+    op.alter_column('users', 'picture_metadata', server_default=None, schema='users')
+    op.create_check_constraint('valid_picture', 'users', "(picture IS NULL) = (picture_metadata::text = 'null')",
+                               schema='users')
+
+
+def downgrade():
+    op.drop_constraint('ck_users_valid_picture', 'users', schema='users')
```
[the remark this answers is not in the document]
Comment (OmeGak) [reply indico/migrations/versions/20200428_1807_532f0ea25bb1_add_user_profile_picture.py:31]: That's correct.

File: indico/modules/users/controllers.py (hunk 1 above)
Comment (OmeGak) [reply indico/modules/users/controllers.py]: Agreed!

**Comment by ThiefMaster** [comment]
When testing with my [octocat avatar](https://user-images.githubusercontent.com/179599/81169165-ac7e9800-8f98-11ea-8eb8-06c2e987547b.png) it somehow got [completely destroyed](https://user-images.githubusercontent.com/179599/81169217-c6b87600-8f98-11ea-97b5-527748648c73.png) during the conversion. Not sure why this happened though... 

It also looks a bit strange right now if the picture is smaller than the circle where it's displayed.

**Comment by ThiefMaster** [comment]
Regarding the upload UI, would it make sense to have a separate box for it on the page instead of including it in the "Details" box?

**Comment by OmeGak** [comment]
> When testing with my [octocat avatar](https://user-images.githubusercontent.com/179599/81169165-ac7e9800-8f98-11ea-8eb8-06c2e987547b.png) it somehow got [completely destroyed](https://user-images.githubusercontent.com/179599/81169217-c6b87600-8f98-11ea-97b5-527748648c73.png) during the conversion. Not sure why this happened though...

I'll look into this. Probably something that I need to pass to `PIL` to preserve the transparency.

> Regarding the upload UI, would it make sense to have a separate box for it on the page instead of including it in the "Details" box?

I've tried a few variations, but I'm not crazy about any of them.

||label|no label|
|-|-|-|
|top|Screen Shot 2020-05-06 at 13 31 44|Screen Shot 2020-05-06 at 13 31 33|
|bottom|Screen Shot 2020-05-06 at 13 35 19|Screen Shot 2020-05-06 at 13 35 12|

**Comment by ThiefMaster** [comment]
True. Doesn't look super amazing. Any other ideas here @indico/development-team?

**Comment by panagiotappl** [comment]
I personally like both the top label and bottom label ideas. IMO, they look good. I would probably expect to see that on top rather than the bottom part of the page.

**Comment by OmeGak** [comment]
> I personally like both the top label and bottom label ideas. IMO, they look good. I would probably expect to see that on top rather than the bottom part of the page.

Two concerns:
- What would you do with the `Block User` button? I would not leave it on the bottom box, but putting it on the `Profile Picture` box looks semantically wrong.
- The label looks redundant, as the box has the same title. We could change the label or box title to something else, but at that point we'll be contorting the UI just for the sake of having two boxes.

Personally, I would just leave it as it is, with a single box.

**Comment by panagiotappl** [comment]
Maybe it's better to leave it in the same box after all. I don't think it looks bad if it's in the same box anyway. I don't see why the picture cannot be in basic details.

**Comment by panagiotappl** [comment]
I'm sorry. I accidentally clicked "Close and comment" . :(

**Comment by OmeGak** [comment]
@ThiefMaster, upload of transparent images is now working fine. I was removing the alpha channel. This fixes it:

```diff
- if pic.mode != 'RGB':
+ if pic.mode not in {'RGB', 'RGBA'}:
     pic = pic.convert('RGB')
```

**Comment by plourenco** [comment]
I like the top options, I'm not sure why but it seems straightforward to have a picture as one of the first details to be edited.

Apart from that, perhaps I could suggest having the image squared (side-by-side) with the details inputs? The same way we're now using in #4408.

File: indico/modules/users/controllers.py [hunk 10]
```diff
@@ -177,13 +183,63 @@ class RHPersonalData(RHUserBase):
...
+        try:
+            pic = Image.open(f)
+        except IOError:
+            raise UserValueError(_('You cannot upload this file as profile picture.'))
+        if pic.format.lower() not in {'jpeg', 'png', 'gif'}:
+            raise UserValueError(_('The file has an invalid format ({format}).').format(format=pic.format))
+        if pic.mode not in {'RGB', 'RGBA'}:
+            pic = pic.convert('RGB')
+        image_bytes = BytesIO()
+        pic = square(pic)
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:220]: `make_square`?

Comment (ThiefMaster) [inline indico/modules/users/controllers.py:220]: I think we also need to resize the image to a max size (256x256px or something like that).

I just managed to upload an 8mb photo of a very pretty cat, but as much as I like hi-res cat pictures, a user picture/avatar is not the place for them ;)

Comment (OmeGak) [reply indico/modules/users/controllers.py:220]: `square` is a also a verb.

**Pull request merged**